### ISSUE TITLE
[BUG] Autologging not detecting metrics that uses make_scorer (GridSearchCV)

### ISSUE THREAD

**Issue by shaikmoeed** [mlflow/mlflow#7358]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 1.29.0

### System information

- **OS Platform and Distribution **: Linux Ubuntu 20.04
- **Python version**: 3.9.13

### Describe the problem

`mlflow.lightgbm.autolog()` is not logging the custom metrics that passed through `make_scorer`. Is there a way or make it possible to use `.autolog()` to detect this metrics?

### Code to reproduce issue

```
from sklearn.datasets import make_classification
from sklearn.model_selection import GridSearchCV
import lightgbm as lgb
from sklearn.metrics import make_scorer
import pandas as pd
import numpy as np

import mlflow

X, y = make_classification()
def my_custom_loss_func_est(estimator, X, y):
    # predictions must be probabilities - e.g. model.predict_proba()
    # example code here:
    diff = np.abs(y - estimator.predict_proba(X)[:, 1]).max()
    return -np.log(1 + diff)

def my_custom_loss_func(ground_truth, predictions):
    # predictions must be probabilities - e.g. model.predict_proba()
    # example code here:
    diff = np.abs(ground_truth - predictions[:, 1]).max()
    return np.log(1 + diff)

mlflow.lightgbm.autolog()
custom_scorer = make_scorer(my_custom_loss_func, 
                            greater_is_better=False,
                            needs_proba=True)

param_grid = {'bossting_type': ["gbdt", "dart"]}
grid = GridSearchCV(lgb.LGBMClassifier(), param_grid=param_grid,
                scoring=custom_scorer, return_train_score=True)
grid.fit(X, y)
pd.DataFrame(grid.cv_results_)[['mean_test_score',
                                   'mean_train_score',
                                   'param_min_samples_leaf',
                                   'param_n_estimators']]
```

### Stack trace

![image](https://user-images.githubusercontent.com/24523726/202367262-292b52b4-f5c5-472a-9982-626f606b3c06.png)

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

**Comment by dbczumar** [comment]
Hi @shaikmoeed, thank you for raising this issue. We would be thrilled to review / merge a PR that implements autologging support for scikit-learn metrics defined with the `make_scorer` API. Feel free to submit one, and we'll work with you to get it merged. Please let me know if you have any questions! Thank you for using MLflow!

**Comment by shaikmoeed** [comment]
@dbczumar Can you please share where (related files/functions locations) to find the related information? I will try to understand the code and share my approach to support this feature (and get verified before creating a PR)

**Comment by dbczumar** [comment]
Hi @shaikmoeed , here's the relevant code entrypoint that MLflow autologging uses to monkey patch scikit-learn metric computations: https://github.com/mlflow/mlflow/blob/589ee6ecbb89680ce718e026e84900d04307f6c0/mlflow/sklearn/__init__.py#L1750-L1756. Please let me know if you have additional questions - happy to help!

**Comment by shaikmoeed** [comment]
@dbczumar  Thanks for directing me to the right files.

I see that, `make_scorer` is getting excluded here https://github.com/mlflow/mlflow/blob/589ee6ecbb89680ce718e026e84900d04307f6c0/mlflow/sklearn/__init__.py#L885

and 

https://github.com/mlflow/mlflow/blob/589ee6ecbb89680ce718e026e84900d04307f6c0/mlflow/sklearn/__init__.py#L1755

are used. By including `make_scorer` and removing  `sklearn.metrics.SCORERS` is enough to detect all the metrics using `make_scorer`?

Still, I was not able to find where exactly the `.autolog` detect metrics and their values while training/tuning parameters 😐

**Comment by dbczumar** [comment]
> @dbczumar Thanks for directing me to the right files.
> 
> I see that, `make_scorer` is getting excluded here
> 
> https://github.com/mlflow/mlflow/blob/589ee6ecbb89680ce718e026e84900d04307f6c0/mlflow/sklearn/__init__.py#L885
> 
> and
> 
> https://github.com/mlflow/mlflow/blob/589ee6ecbb89680ce718e026e84900d04307f6c0/mlflow/sklearn/__init__.py#L1755
> 
> are used. By including `make_scorer` and removing `sklearn.metrics.SCORERS` is enough to detect all the metrics using `make_scorer`?
> 
> Still, I was not able to find where exactly the `.autolog` detect metrics and their values while training/tuning parameters 😐

I think we'll need to create a separate monkey patch for `make_scorer` using MLflow's `safe_patch` utility. This patched `make_scorer` should also call `safe_patch` to create & return a monkey patched `scorer`. This returned scorer should log metrics to MLflow when the scorer is called.

**Comment by dbczumar** [comment]
Hi @shaikmoeed , any updates here?

**Comment by shaikmoeed** [comment]
@dbczumar Due to other priorities, I didn't get a chance to look into this further. I will get some free time in a week, to continue on this FR.

**Comment by shaikmoeed** [comment]
Hi @dbczumar, there is some discrepancy in our understanding. I was expecting to use `mlflow.lightgbm.autolog()` to log all the custom/user-defined metrics(functions) wrapped inside `make_scorer` instead of manually logging all the custom metrics using `mlflow.log_metrics(metrics)`. Is that what you are expecting from this FR too? If yes, what I want to know is as follows:

1. How to collect all the metrics that are wrapped inside `make_scorer`?
2. What should be named for this custom metric? (can it be `custom_metric_function_name`?)
3. What are the prerequisites(codes, test cases, concepts specific to mlflow) that I need to be aware of before starting this task?

**Comment by dbczumar** [comment]
> Hi @dbczumar, there is some discrepancy in our understanding. I was expecting to use `mlflow.lightgbm.autolog()` to log all the custom/user-defined metrics(functions) wrapped inside `make_scorer` instead of manually logging all the custom metrics using `mlflow.log_metrics(metrics)`. Is that what you are expecting from this FR too? If yes, what I want to know is as follows:
> 
> 1. How to collect all the metrics that are wrapped inside `make_scorer`?
> 2. What should be named for this custom metric? (can it be `custom_metric_function_name`?)
> 3. What are the prerequisites(codes, test cases, concepts specific to mlflow) that I need to be aware of before starting this task?

@shaikmoeed Good questions!

1. We don't need to collect these metrics, we just need to monkey patch `make_scorer` so that it returns a scorer with a custom implementation that contains `log_metric()` calls.
2. The metric name should be the name of the `score_func` passed to https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html
3. Monkey patching is the most important concept. https://github.com/mlflow/mlflow/issues/7358#issuecomment-1329870914 should be a good starting resource, but let me know if you have specific questions :)

**Comment by shaikmoeed** [comment]
Hi @dbczumar 

I have created monkey patched `make_scorer` at a high level:
```
def patched_make_scorer(original, *args, **kwargs):
    # Call the original make_scorer function to get the unpatched scorer
    scorer = original(*args, **kwargs)

    # Define the patch function to log metrics to MLflow when the scorer is called
    def log_metrics(scorer, *args, **kwargs):
        # Call the original scorer function to get the score
        metric_value= scorer(*args, **kwargs)

        # Log the score as a metric to MLflow
        mlflow.log_metric("score", metric_value)

        # Return the score
        return metric_value

    # Use safe_patch to create and return a monkey patched scorer
    patched_scorer = safe_patch(scorer, "score", log_metrics, manage_run=False)
    return patched_scorer
```

Finally, I can use `safe_patch` to create the patched `make_scorer` function:
```
safe_patch(make_scorer, sklearn.metrics, "make_scorer", patched_make_scorer, manage_run=False)
```

Is that correct? Your thoughts on this code? Am I missing anything?

Can I get `y_true`, `y_pred` values in `**kwargs`?
